python/ray/data/dataset_iterator.py

```diff
@@ -0,0 +1,248 @@
+import abc
+from typing import TYPE_CHECKING, Dict, List, Optional, Union, Iterator
+from typing_extensions import Literal
+
+from ray.data.block import DataBatch
+from ray.util.annotations import PublicAPI
+
+if TYPE_CHECKING:
+    import tensorflow as tf
+    import torch
+    from ray.data._internal.torch_iterable_dataset import TorchTensorBatchType
+
+
+@PublicAPI(stability="beta")
+class DatasetIterator(abc.ABC):
+    """An iterator for reading items from a :class:`~Dataset` or
+    :class:`~DatasetPipeline`.
+
+    For Datasets, each iteration call represents a complete read of all items in the
+    Dataset. For DatasetPipelines, each iteration call represents one pass (epoch)
+    over the Dataset. Note that for DatasetPipelines, each pass iterates over
+    the original Dataset, instead of a window (if `.window()` was used).
```

Comment:
Nit: Inlined code that isn't a cross-reference link should be double-backticked.
```suggestion
    the original Dataset, instead of a window (if ``.window()`` was used).
```